ndb/src/google/cloud/ndb/model.py

```diff
@@ -2538,11 +2538,111 @@ def _now():
+
+    This is similar to a mix-in, for :class:`StructuredProperty` and
+    :class:`LocalStructuredProperty`.
+
+    This converts sub-entities to dictionaries by calling ``to_dict()`` on them
+    which allows nesting to go arbitrarily deep.
+
+    Args:
+        property_ (Property): Usually ``self`` in the caller.
+        entity (Model): An entity to get a value from.
+
+    Returns:
+        Optional[Union[dict, List[dict]]]: The nested dictionary (or list of
+        dictionaries in the repeated case) stored for ``property_``.
+    """
+    value = property_._get_value(entity)
+    if property_._repeated:
+        value = [sub_value._to_dict() for sub_value in value]
+    elif value is not None:
+        value = value._to_dict()
+
+    return value
+
+
 class StructuredProperty(Property):
-    __slots__ = ()
+    """A property that contains a model entity as value.
 
-    def __init__(self, *args, **kwargs):
-        raise NotImplementedError
+    If this property is indexed, the values of any sub-entities are also
+    indexed and can be queried.
+
+    Args:
+        modelclass (type): A subclass of :class:`Model`. This defines the
+            type of entity that will be stored for this property.
+        name (str): The name of the property.
+        indexed (bool): Indicates if the value should be indexed.
+        repeated (bool): Indicates if this property is repeated, i.e. contains
+            multiple values.
+        required (bool): Indicates if this property is required on the given
+            model type.
+        default (Any): The default value for this property.
+        choices (Iterable[Any]): A container of allowed values for this
+            property.
+        validator (Callable[[~google.cloud.ndb.model.Property, Any], bool]): A
+            validator to be used to check values.
+        verbose_name (str): A longer, user-friendly name for this property.
+        write_empty_list (bool): Indicates if an empty list should be written
+            to the datastore.
+
+    Raises:
+        TypeError: If ``repeated=True`` but the ``modelclass`` has
+            repeated properties or subproperties.
+    """
+
+    _modelclass = None
+
+    def __init__(
+        self,
+        modelclass,
+        name=None,
+        *,
+        compressed=None,
+        indexed=None,
+        repeated=None,
+        required=None,
+        default=None,
+        choices=None,
+        validator=None,
+        verbose_name=None,
+        write_empty_list=None
+    ):
+        super(StructuredProperty, self).__init__(
+            name=name,
+            indexed=indexed,
+            repeated=repeated,
+            required=required,
+            default=default,
+            choices=choices,
+            validator=validator,
+            verbose_name=verbose_name,
+            write_empty_list=write_empty_list,
+        )
+        if self._repeated and modelclass._has_repeated:
+            raise TypeError(
```

Comment:
This wouldn't be necessary with the new API. However, the backend may still enforce it, but the protobuf does not (since it can just use a nested `Entity` as this property).